---
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{motifmatchr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# motifmatchr

The motifmatchr package is designed for analyzing many sequences and many motifs to find which sequences contain which motifs. It uses the MOODS C++ library (developedby Pasi Rastas, Janne Korhonen, and Petri Martinmaki) internally for motif matching.  The primary method of motifmatchr is `match_motifs`, which takes in motif PWMs/PFMs and genomic ranges or sequences and returns either which ranges/sequences match which motifs or the positions of the matches.  

Compared with alternative motif matching functions available in Bioconductor (e.g. matchPWM in Biostrings or searchSeq in TFBSTools), motifmatchr is designed specifically for the use case of determining whether many different sequences/ranges contain many different motifs. 

## Installation

Installation is easiest using the devtools package. The function `install_github` will install the package.

``` r
devtools::install_github("GreenleafLab/motifmatchr", auth_token = "my_token")
```

The argument auth\_token takes in your github [personal access token](https://github.com/settings/tokens). This token is needed because at the moment this repository is private.

A number of needed packages are installed in this process. Depending on your repository settings, the Bioconductor dependencies may fail to install. Use `setRepositories(graphics=FALSE)` to see what repositories you have activated and to add the BioC software repository if need be.

## Quickstart

Example use case of motifmatchr with a set of peaks and a few motifs. For additional options for inputs & outputs, see remainder of vignette.

```{r, message = FALSE}
library(motifmatchr)
library(GenomicRanges)
library(SummarizedExperiment)

# load some example motifs
data(example_motifs, package = "motifmatchr") 

# Make a set of peaks
peaks <- GRanges(seqnames = c("chr1","chr2","chr2"),
                 ranges = IRanges(start = c(76585873,42772928,100183786),
                                  width = 500))

# Get motif matches for example motifs in peaks (using default hg19 genome)
motif_ix <- match_motifs(example_motifs, peaks) 
motif_matches(motif_ix) # Extract matches matrix from result

# Get motif positions within peaks for example motifs in peaks 
motif_pos <- match_motifs(example_motifs, peaks, out = "positions") 
```

# Inputs

This method has two mandatory arguments:

1) Position weight matrices or position frequency matrices, stored in the PWMatrix, PFMatrix, PWMatrixList, or PFMatrixList objects from the TFBSTools package

2) Either a set of genomic ranges (GenomicRanges or RangedSummarizedExperiment object) or a set of sequences (either DNAStringSet, DNAString, or simple character vector)

If the second argument is a set of genomic ranges, a genome sequence is also required. By default [BSgenome.Hsapiens.UCSC.hg19](https://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html) is used. If using another genome build, the appropraiate BSgenome object for your species should be passed to the `genome` argument.

```{r}
# using peaks
motif_ix_peaks <- match_motifs(example_motifs, peaks) 

# using SummarizedExperiment
example_SummarizedExperiment <- 
    SummarizedExperiment(assays = list(counts = matrix(1,
                                                       ncol = 4,
                                                       nrow = 3)),
                         rowRanges = peaks)

motif_ix_SummarizedExperiment <- match_motifs(example_motifs,
                                              example_SummarizedExperiment) 

all.equal(motif_matches(motif_ix_peaks),
          motif_matches(motif_ix_SummarizedExperiment))
```

```{r, message = FALSE}
# using DNAStringSet
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg19)

example_DNAStringSet <- getSeq(BSgenome.Hsapiens.UCSC.hg19,
                          peaks)

motif_ix_DNAStringSet <- match_motifs(example_motifs, example_DNAStringSet) 

all.equal(motif_matches(motif_ix_peaks), motif_matches(motif_ix_DNAStringSet))
```


```{r, message = FALSE}
# using character vector
example_character <- as.character(example_DNAStringSet)

motif_ix_character <- match_motifs(example_motifs, example_character) 


all.equal(motif_matches(motif_ix_peaks), motif_matches(motif_ix_character))
```


## Options

### Background nucleotide frequencies

In determining motif matches, background nucleotide frequencies are used.  By default the background frequencies are the nucleotide frequencies within the input sequence -- to use alternate frequencies, supply the `bg` arument to `match_pwms`.  If the input sequences are fairly short (as in our vignette example!), it probably makes sense to use other input frequencies.  Here we show how to use even frequencies:

```{r}
motif_ix <- match_motifs(example_motifs, peaks, bg = rep(0.25, 4)) 
```

To specify frequencies per base pair, the order should be "A","C","G", then "T", or those nucleotides should be used as names in the vector.

```{r}
motif_ix <- match_motifs(example_motifs, peaks, 
                         bg = c("A" = 0.2,"C" = 0.3, "G" = 0.3, "T" = 0.2)) 
```

PWMatrix objects have associated background frequencies that can be accessed using the `bg` function. If the supplied background frequencies to `match_pwms` do not match the frequencies in the input PWM, then the PWM is adjusted to refect the supplied background frequencies. The calculated score is based on this adjusted PWM and not the direct input PWM. To ensure the score is computed using the exact PWM input, simply make sure the background frequencies passed to match_motifs match hose used for the input PWM and stored in bg slot of PWMatrix object.

### Log base and pseudocounts

motifmatchr expects input PWMs to use either natural logarithm or log 2. If the input is a PFM, the TFBSTools toPWM is used for making the PWM with the default psueodcounts of 0.8 and base 2 logarithm. For more control of the pseudocounts, simply use the toPWM function to convert your PFMs prior to calling match_motifs. 

```{r, message = FALSE}
library(TFBSTools)
example_pwms <- do.call(PWMatrixList,lapply(example_motifs, toPWM, 
                                            pseudocounts = 0.5))
```

### P value

The default p-value cutoff is 0.00005. No adjustment for multiple comparisons is made in this p-value cutoff. This p-value threshold is used to determine a score threshold. 


# Outputs

The match_motifs can return three possible outputs, depending on the `out` argument:

1) (Default, with `out = matches`) Boolean matrix indicating which ranges/sequences contain which motifs, stored as "matches" in assays slot of SummarizedExperiment object

2) (out = `scores`) Same as (1) plus two additional assays -- a matrix with the score of the high motif score within each range/sequence (score only reported if match present) and a matrix with the number of motif matches.

3) (out = `positions`) A GenomicRangesList with the ranges of all matches within the input ranges/sequences. 



```{r}
motif_ix <- match_motifs(example_motifs, peaks) 
print(motif_ix)

motif_ix_scores <- match_motifs(example_motifs, peaks, out = "scores")
print(motif_ix_scores)

motif_pos <- match_motifs(example_motifs, peaks, out = "positions") 
print(motif_pos)
```

## Session Info

```{r}
Sys.Date()
```

```{r}
sessionInfo()
```



