<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title> &bull; motifmatchr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">motifmatchr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/motifmatchr.html">Vignette</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1></h1>
            
          </div>

    
    
<div class="contents">
<div id="motifmatchr" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#motifmatchr" class="anchor"> </a></body></html>motifmatchr</h1>
<p>The motifmatchr package is designed for analyzing many sequences and many motifs to find which sequences contain which motifs. It uses the MOODS C++ library internally for motif matching. The primary method of motifmatchr is <code>match_motifs</code>, which takes in motif PWMs/PFMs and genomic ranges or sequences and returns either which ranges/sequences match which motifs or the positions of the matches.</p>
<p>Compared with alternative motif matching functions available in Bioconductor (e.g.&nbsp;matchPWM in Biostrings or searchSeq in TFBSTools), motifmatchr is designed specifically for the use case of determining whether many different sequences/ranges contain many different motifs.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#installation" class="anchor"> </a></body></html>Installation</h2>
<p>Installation is easiest using the devtools package. The function <code>install_github</code> will install the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">"GreenleafLab/motifmatchr"</span>, <span class="dt">auth_token =</span> <span class="st">"my_token"</span>)</code></pre></div>
<p>The argument auth_token takes in your github <a href="https://github.com/settings/tokens">personal access token</a>. This token is needed because at the moment this repository is private.</p>
<p>A number of needed packages are installed in this process. Depending on your repository settings, the Bioconductor dependencies may fail to install. Use <code>setRepositories(graphics=FALSE)</code> to see what repositories you have activated and to add the BioC software repository if need be.</p>
</div>
<div id="quickstart" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#quickstart" class="anchor"> </a></body></html>Quickstart</h2>
<p>Example use case of motifmatchr with a set of peaks and a few motifs. For additional options for inputs &amp; outputs, see remainder of vignette.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(GenomicRanges)
<span class="kw">library</span>(SummarizedExperiment)

<span class="co"># load some example motifs</span>
<span class="kw">data</span>(example_motifs, <span class="dt">package =</span> <span class="st">"motifmatchr"</span>) 

<span class="co"># Make a set of peaks</span>
peaks &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">"chr1"</span>,<span class="st">"chr2"</span>,<span class="st">"chr2"</span>),
                 <span class="dt">ranges =</span> <span class="kw">IRanges</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">76585873</span>,<span class="dv">42772928</span>,<span class="dv">100183786</span>),
                                  <span class="dt">width =</span> <span class="dv">500</span>))

<span class="co"># Get motif matches for example motifs in peaks (using hg19 genome, the default)</span>
motif_ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks) 
<span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix) <span class="co"># Extract matches matrix from SummarizedExperiment result</span></code></pre></div>
<pre><code>## 3 x 3 sparse Matrix of class "lgCMatrix"
##      MA0599.1_KLF5 MA0107.1_RELA MA0137.3_STAT1
## [1,]             |             |              .
## [2,]             |             .              |
## [3,]             |             .              |</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get motif positions within peaks for example motifs in peaks </span>
motif_pos &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks, <span class="dt">out =</span> <span class="st">"positions"</span>) </code></pre></div>
</div>
</div>
<div id="inputs" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#inputs" class="anchor"> </a></body></html>Inputs</h1>
<p>This method has two mandatory arguments:</p>
<ol style="list-style-type: decimal"><li><p>Position weight matrices or position frequency matrices, stored in the PWMatrix, PFMatrix, PWMatrixList, or PFMatrixList objects from the TFBSTools package</p></li>
<li><p>Either a set of genomic ranges (GenomicRanges or RangedSummarizedExperiment object) or a set of sequences (either DNAStringSet, DNAString, or simple character vector)</p></li>
</ol><p>If the second argument is a set of genomic ranges, a genome sequence is also required. By default <a href="https://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html">BSgenome.Hsapiens.UCSC.hg19</a> is used. If using another genome build, the appropraiate BSgenome object for your species should be passed to the <code>genome</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using peaks</span>
motif_ix_peaks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks) 

<span class="co"># using SummarizedExperiment</span>
example_SummarizedExperiment &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">matrix</span>(<span class="dv">1</span>,
                                                                               <span class="dt">ncol =</span> <span class="dv">4</span>,
                                                                               <span class="dt">nrow =</span> <span class="dv">3</span>)),
                                                 <span class="dt">rowRanges =</span> peaks)

motif_ix_SummarizedExperiment &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, example_SummarizedExperiment) 

<span class="kw">all.equal</span>(<span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_peaks), <span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_SummarizedExperiment))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using DNAStringSet</span>
<span class="kw">library</span>(Biostrings)
<span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)

example_DNAStringSet &lt;-<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19,
                          peaks)

motif_ix_DNAStringSet &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, example_DNAStringSet) 

<span class="kw">all.equal</span>(<span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_peaks), <span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_DNAStringSet))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using character vector</span>
example_character &lt;-<span class="st"> </span><span class="kw">as.character</span>(example_DNAStringSet)

motif_ix_character &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, example_character) 


<span class="kw">all.equal</span>(<span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_peaks), <span class="kw"><a href="../reference/motif_matches.html">motif_matches</a></span>(motif_ix_character))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="options" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#options" class="anchor"> </a></body></html>Options</h2>
<div id="background-nucleotide-frequencies" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#background-nucleotide-frequencies" class="anchor"> </a></body></html>Background nucleotide frequencies</h3>
<p>In determining motif matches, background nucleotide frequencies are used. By default the background frequencies are the nucleotide frequencies within the input sequence &ndash; to use alternate frequencies, supply the <code>bg</code> arument to <code>match_pwms</code>. If the input sequences are fairly short (as in our vignette example!), it probably makes sense to use other input frequencies. Here we show how to use even frequencies:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motif_ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks, <span class="dt">bg =</span> <span class="kw">rep</span>(<span class="fl">0.25</span>, <span class="dv">4</span>)) </code></pre></div>
<p>To specify frequencies per base pair, the order should be &ldquo;A&rdquo;,&ldquo;C&rdquo;,&ldquo;G&rdquo;, then &ldquo;T&rdquo;, or those nucleotides should be used as names in the vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motif_ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks, <span class="dt">bg =</span> <span class="kw">c</span>(<span class="st">"A"</span> =<span class="st"> </span><span class="fl">0.2</span>,<span class="st">"C"</span> =<span class="st"> </span><span class="fl">0.3</span>, <span class="st">"G"</span> =<span class="st"> </span><span class="fl">0.3</span>, <span class="st">"T"</span> =<span class="st"> </span><span class="fl">0.2</span>)) </code></pre></div>
<p>PWMatrix objects have associated background frequencies that can be accessed using the <code>bg</code> function. If the supplied background frequencies to <code>match_pwms</code> do not match the frequencies in the input PWM, then the PWM is adjusted to refect the supplied background frequencies. The calculated score is based on this adjusted PWM and not the direct input PWM. To ensure the score is computed using the exact PWM input, simply make sure the background frequencies passed to match_motifs match hose used for the input PWM and stored in bg slot of PWMatrix object.</p>
</div>
<div id="log-base-and-pseudocounts" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#log-base-and-pseudocounts" class="anchor"> </a></body></html>Log base and pseudocounts</h3>
<p>motifmatchr expects input PWMs to use either natural logarithm or log 2. If the input is a PFM, the TFBSTools toPWM is used for making the PWM with the default psueodcounts of 0.8 and base 2 logarithm. For more control of the pseudocounts, simply use the toPWM function to convert your PFMs prior to calling match_motifs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TFBSTools)
example_pwms &lt;-<span class="st"> </span><span class="kw">do.call</span>(PWMatrixList,<span class="kw">lapply</span>(example_motifs, toPWM, <span class="dt">pseudocounts =</span> <span class="fl">0.5</span>))</code></pre></div>
</div>
<div id="p-value" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#p-value" class="anchor"> </a></body></html>P value</h3>
<p>The default p-value cutoff is 0.00005. No adjustment for multiple comparisons is made in this p-value cutoff. This p-value threshold is used to determine a score threshold.</p>
</div>
</div>
</div>
<div id="outputs" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#outputs" class="anchor"> </a></body></html>Outputs</h1>
<p>The match_motifs can return three possible outputs, depending on the <code>out</code> argument:</p>
<ol style="list-style-type: decimal"><li><p>(Default, with <code>out = matches</code>) Boolean matrix indicating which ranges/sequences contain which motifs, stored as &ldquo;matches&rdquo; in assays slot of SummarizedExperiment object</p></li>
<li><p>(out = <code>scores</code>) Same as (1) plus two additional assays &ndash; a matrix with the score of the high motif score within each range/sequence (score only reported if match present) and a matrix with the number of motif matches.</p></li>
<li><p>(out = <code>positions</code>) A GenomicRangesList with the ranges of all matches within the input ranges/sequences.</p></li>
</ol><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motif_ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks) 
<span class="kw">print</span>(motif_ix)</code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 3 3 
## metadata(0):
## assays(1): matches
## rownames: NULL
## rowData names(0):
## colnames(3): MA0599.1_KLF5 MA0107.1_RELA MA0137.3_STAT1
## colData names(2): pwm name</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motif_ix_scores &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks, <span class="dt">out =</span> <span class="st">"scores"</span>)
<span class="kw">print</span>(motif_ix_scores)</code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 3 3 
## metadata(0):
## assays(3): scores matches counts
## rownames: NULL
## rowData names(0):
## colnames(3): MA0599.1_KLF5 MA0107.1_RELA MA0137.3_STAT1
## colData names(2): pwm name</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motif_pos &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match_motifs.html">match_motifs</a></span>(example_motifs, peaks, <span class="dt">out =</span> <span class="st">"positions"</span>) 
<span class="kw">print</span>(motif_pos)</code></pre></div>
<pre><code>## $MA0599.1_KLF5
## GRanges object with 3 ranges and 1 metadata column:
##       seqnames                 ranges strand |            score
##          &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;numeric&gt;
##   [1]     chr1 [ 76585928,  76585937]      - |  7.9751454917072
##   [2]     chr2 [ 42773248,  42773257]      - | 7.08576191714116
##   [3]     chr2 [100184023, 100184032]      - | 7.73960919591018
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $MA0107.1_RELA
## GRanges object with 1 range and 1 metadata column:
##       seqnames               ranges strand |            score
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;numeric&gt;
##   [1]     chr1 [76585933, 76585942]      + | 13.5563372551788
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $MA0137.3_STAT1
## GRanges object with 3 ranges and 1 metadata column:
##       seqnames                 ranges strand |            score
##          &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;numeric&gt;
##   [1]     chr2 [ 42773087,  42773097]      - | 16.2940093372611
##   [2]     chr2 [100183852, 100183862]      + | 14.8129754070887
##   [3]     chr2 [100183852, 100183862]      - |  13.953566440506
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div id="session-info" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#session-info" class="anchor"> </a></body></html>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-02-02"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.1 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] motifmatchr_0.1.0                 BSgenome.Hsapiens.UCSC.hg19_1.4.0
##  [3] BSgenome_1.42.0                   rtracklayer_1.34.0               
##  [5] SummarizedExperiment_1.4.0        Biobase_2.34.0                   
##  [7] GenomicRanges_1.26.1              GenomeInfoDb_1.10.0              
##  [9] TFBSTools_1.12.0                  Biostrings_2.42.0                
## [11] XVector_0.14.0                    IRanges_2.8.0                    
## [13] S4Vectors_0.12.0                  BiocGenerics_0.20.0              
## 
## loaded via a namespace (and not attached):
##  [1] pkgload_0.0.0.9000          VGAM_1.0-2                 
##  [3] splines_3.3.1               R.utils_2.5.0              
##  [5] gtools_3.5.0                assertthat_0.1             
##  [7] Rsamtools_1.26.1            yaml_2.1.14                
##  [9] DirichletMultinomial_1.16.0 highlight_0.4.7            
## [11] RSQLite_1.1-1               backports_1.0.4            
## [13] lattice_0.20-34             digest_0.6.10              
## [15] colorspace_1.2-7            htmltools_0.3.5            
## [17] Matrix_1.2-7.1              R.oo_1.21.0                
## [19] plyr_1.8.4                  XML_3.98-1.5               
## [21] devtools_1.12.0.9000        zlibbioc_1.20.0            
## [23] purrr_0.2.2                 xtable_1.8-2               
## [25] GO.db_3.4.0                 scales_0.4.1               
## [27] whisker_0.3-2               BiocParallel_1.8.0         
## [29] tibble_1.2                  annotate_1.52.0            
## [31] ggplot2_2.2.0               withr_1.0.2                
## [33] TFMPvalue_0.0.6             lazyeval_0.2.0             
## [35] magrittr_1.5                crayon_1.3.2               
## [37] memoise_1.0.0               poweRlaw_0.60.3            
## [39] evaluate_0.10               R.methodsS3_1.7.1          
## [41] RcppArmadillo_0.7.600.1.0   xml2_1.0.0.9001            
## [43] pkgbuild_0.0.0.9000         CNEr_1.10.0                
## [45] tools_3.3.1                 stringr_1.1.0              
## [47] munsell_0.4.3               callr_1.0.0.9000           
## [49] AnnotationDbi_1.36.0        pkgdown_0.1.0.9000         
## [51] caTools_1.17.1              grid_3.3.1                 
## [53] RCurl_1.95-4.8              bitops_1.0-6               
## [55] rmarkdown_1.3               testthat_1.0.2             
## [57] gtable_0.2.0                DBI_0.5-1                  
## [59] roxygen2_5.0.1              reshape2_1.4.2             
## [61] R6_2.2.0                    GenomicAlignments_1.10.0   
## [63] knitr_1.15.1                seqLogo_1.40.0             
## [65] rprojroot_1.1               readr_1.0.0                
## [67] desc_1.0.1                  stringi_1.1.2              
## [69] Rcpp_0.12.8</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#motifmatchr">motifmatchr</a><ul class="nav nav-pills nav-stacked"><li><a href="#installation">Installation</a></li>
      <li><a href="#quickstart">Quickstart</a></li>
      </ul></li>
      <li><a href="#inputs">Inputs</a><ul class="nav nav-pills nav-stacked"><li><a href="#options">Options</a></li>
      </ul></li>
      <li><a href="#outputs">Outputs</a><ul class="nav nav-pills nav-stacked"><li><a href="#session-info">Session Info</a></li>
      </ul></li>
      </ul></div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

  </body></html>
