<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>match_motifs — match_motifs • motifmatchr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">motifmatchr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/motifmatchr.html">Vignette</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/motifmatchr/">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>match_motifs</h1>
    </div>

    
    <p>Find motif matches</p>
    

    <pre><span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='no'>...</span>)

<span class='co'># S4 method for PWMatrixList,DNAStringSet</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>,
  <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>, <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>,
  <span class='st'>"scores"</span>, <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>, <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for PWMatrixList,character</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>, <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>, <span class='st'>"scores"</span>,
  <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>, <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for PWMatrixList,DNAString</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>, <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>, <span class='st'>"scores"</span>,
  <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>, <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for PWMatrixList,GenomicRanges</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>,
  <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>GenomeInfoDb</span><span class='kw'>::</span><span class='fu'>genome</span>(<span class='no'>subject</span>), <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>,
  <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>, <span class='st'>"scores"</span>, <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>,
  <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>)

<span class='co'># S4 method for PWMatrixList,RangedSummarizedExperiment</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>,
  <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>GenomeInfoDb</span><span class='kw'>::</span><span class='fu'>genome</span>(<span class='no'>subject</span>), <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>,
  <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>, <span class='st'>"scores"</span>, <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>,
  <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>)

<span class='co'># S4 method for PWMatrixList,BSgenomeViews</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>,
  <span class='kw'>bg</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"subject"</span>, <span class='st'>"genome"</span>, <span class='st'>"even"</span>), <span class='kw'>out</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"matches"</span>, <span class='st'>"scores"</span>,
  <span class='st'>"positions"</span>), <span class='kw'>p.cutoff</span> <span class='kw'>=</span> <span class='fl'>5e-05</span>, <span class='kw'>w</span> <span class='kw'>=</span> <span class='fl'>7</span>)

<span class='co'># S4 method for PFMatrixList,ANY</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='no'>...</span>)

<span class='co'># S4 method for PWMatrix,ANY</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='no'>...</span>)

<span class='co'># S4 method for PFMatrix,ANY</span>
<span class='fu'>match_motifs</span>(<span class='no'>pwms</span>, <span class='no'>subject</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>pwms</dt>
      <dd>either <code><a href='http://www.rdocumentation.org/packages/TFBSTools/topics/XMatrix-class'>PFMatrix</a></code>,
<code><a href='http://www.rdocumentation.org/packages/TFBSTools/topics/XMatrixList-class'>PFMatrixList</a></code>, <code><a href='http://www.rdocumentation.org/packages/TFBSTools/topics/XMatrix-class'>PWMatrix</a></code>,
<code><a href='http://www.rdocumentation.org/packages/TFBSTools/topics/XMatrixList-class'>PWMatrixList</a></code></dd>
      <dt>subject</dt>
      <dd>either <code><a href='http://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GenomicRanges</a></code>,
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-class'>DNAStringSet</a></code>, <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></code>,
or character vector</dd>
      <dt>...</dt>
      <dd>additional arguments depending on inputs</dd>
      <dt>genome</dt>
      <dd>BSgenome object, <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-class'>DNAStringSet</a></code>, or
<code><a href='http://www.rdocumentation.org/packages/Rsamtools/topics/FaFile-class'>FaFile</a></code>, or short string signifying genome build
recognized by <code><a href='http://www.rdocumentation.org/packages/BSgenome/topics/available.genomes'>getBSgenome</a></code>. Only required if
subject is <code><a href='http://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GenomicRanges</a></code> or
<code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>RangedSummarizedExperiment</a></code> or if bg is set
to &quot;genome&quot;</dd>
      <dt>bg</dt>
      <dd>background nucleotide frequencies. Default is to compute based on
subject, i.e. the specific set of sequences being evaluated. See Details.</dd>
      <dt>out</dt>
      <dd>what to return? see return section</dd>
      <dt>p.cutoff</dt>
      <dd>p-value cutoff for returning motifs</dd>
      <dt>w</dt>
      <dd>parameter controlling size of window for filtration; default is 7</dd>
      <dt>ranges</dt>
      <dd>if subject is not GenomicRanges or RangedSummarizedExperiment,
these ranges can be used to specify what ranges the input sequences
correspond to. These ranges will be incorporated into the
SummarizedExperiment output if out is &quot;matches&quot; or &quot;scores&quot; or will be used
to give absolute positions of motifs if out is &quot;positions&quot;</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Either returns a SummarizedExperiment with a sparse matrix with
 values set to TRUE for a match (if out == &#39;matches&#39;), a
 SummarizedExperiment with a matches matrix as well as matrices with the
 maximum motif score and total motif counts (if out == &#39;scores&#39;), or a
 <code><a href='http://www.rdocumentation.org/packages/GenomicRanges/topics/GenomicRangesList-class'>GenomicRangesList</a></code> or a list of
 <code><a href='http://www.rdocumentation.org/packages/IRanges/topics/IRangesList-class'>IRangesList</a></code> with all the positions of matches
 (if out == &#39;positions&#39;)</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Background nucleotide frequencies can be set to &quot;subject&quot; to use the
subject sequences or ranges for computing the nucleotide frequencies,
&quot;genome&quot; for using the genomice frequencies (in which case a genome must be
specified), &quot;even&quot; for using 0.25 for each base, or a numeric vector with A,
C, G, and T frequencies.</p>
    
    <h2 class="hasAnchor" id="methods-by-class-"><a class="anchor" href="#methods-by-class-"></a>Methods (by class)</h2>

    <p><ul>
<li> <code>pwms = PWMatrixList,subject = DNAStringSet</code>: PWMatrixList/DNAStringSet</p>
    <p></li>
<li> <code>pwms = PWMatrixList,subject = character</code>: PWMatrixList/character</p>
    <p></li>
<li> <code>pwms = PWMatrixList,subject = DNAString</code>: PWMatrixList/DNAString</p>
    <p></li>
<li> <code>pwms = PWMatrixList,subject = GenomicRanges</code>: PWMatrixList/GenomicRanges</p>
    <p></li>
<li> <code>pwms = PWMatrixList,subject = RangedSummarizedExperiment</code>: PWMatrixList/RangedSummarizedExperiment</p>
    <p></li>
<li> <code>pwms = PWMatrixList,subject = BSgenomeViews</code>: PWMatrixList/BSGenomeViews</p>
    <p></li>
<li> <code>pwms = PFMatrixList,subject = ANY</code>: PFMatrixList/ANY</p>
    <p></li>
<li> <code>pwms = PWMatrix,subject = ANY</code>: PWMatrix/ANY</p>
    <p></li>
<li> <code>pwms = PFMatrix,subject = ANY</code>: PFMatrix/ANY
</li>
</ul></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'>data</span>(<span class='no'>example_motifs</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"motifmatchr"</span>)

<span class='co'># Make a set of peaks</span>
<span class='no'>peaks</span> <span class='kw'>&lt;-</span> <span class='kw'>GenomicRanges</span><span class='kw'>::</span><span class='fu'>GRanges</span>(<span class='kw'>seqnames</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"chr1"</span>,<span class='st'>"chr2"</span>,<span class='st'>"chr2"</span>),
                <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='kw'>IRanges</span><span class='kw'>::</span><span class='fu'>IRanges</span>(<span class='kw'>start</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>76585873</span>,<span class='fl'>42772928</span>,
                                          <span class='fl'>100183786</span>),
                                          <span class='kw'>width</span> <span class='kw'>=</span> <span class='fl'>500</span>))

<span class='co'># Get motif matches for example motifs</span>
<span class='no'>motif_ix</span> <span class='kw'>&lt;-</span> <span class='fu'>match_motifs</span>(<span class='no'>example_motifs</span>, <span class='no'>peaks</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#methods-by-class-">Methods (by class)</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://www.aliciaschep.com'>Alicia Schep</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
